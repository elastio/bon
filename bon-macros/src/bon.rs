use crate::builder;
use crate::normalization::{ExpandCfg, ExpansionOutput};
use crate::util::prelude::*;

pub(crate) fn generate(params: TokenStream2, item: TokenStream2) -> TokenStream2 {
    try_generate(params, item.clone())
        .unwrap_or_else(|err| crate::error::error_into_token_stream(err, item))
}

pub(crate) fn try_generate(params: TokenStream2, item: TokenStream2) -> Result<TokenStream2> {
    let item: syn::Item = syn::parse2(item)?;
    let macro_path = syn::parse_quote!(::bon::bon);

    let ctx = ExpandCfg {
        macro_path,
        params,
        item,
    };

    let (_params, item) = match ctx.expand_cfg()? {
        ExpansionOutput::Expanded { params, item } => (params, item),
        ExpansionOutput::Recurse(output) => return Ok(output),
    };

    match item {
        syn::Item::Impl(item_impl) => builder::item_impl::generate(item_impl),
        _ => bail!(
            &item,
            "`#[bon]` attribute is expected to be placed on an `impl` block \
             but it was placed on other syntax instead"
        ),
    }
}
