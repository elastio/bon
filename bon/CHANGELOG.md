# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1](https://github.com/elastio/bon/compare/bon-v0.1.0...bon-v0.1.1) - 2024-07-23

### Other
- Upd docs
- More improvements, deploy-website workflow
- Docs for alternatives plus `#[builder(name = ...)]` for members
- Rename `field -> member`
- Add more docs
- Add doctests, more docs
- More docs
- More docs and better decisions
- Fix builder attrs not being removed correctly
- Next iteration
- Test struct input
- Start testing struct input
- Refactor in preparation for struct input for macro
- Make `#[default]` compatible with `Option`
- Add `default` and `into` attrs
- Make bool optional, strip `_` prefix
- Add special casing for `Option` setters
- Better doc comments
- Improve nested refs handling
- Cleanup
- Fix a bunch of papercuts with codegen
- Finish implementing support for associated functions
- Start adding support for associated functions
- Add some docs
- Rebranding buildy -> prox (buildy crate name is already taken on crates.io)
