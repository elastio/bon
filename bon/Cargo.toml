[package]
name    = "bon"
version = "2.3.0"

description = "Next-gen compile-time-checked builder generator, named function's arguments, and more!"

categories = [
  "rust-patterns",
  "data-structures",
  "asynchronous",
  "no-std",
  "no-std::no-alloc",
]
keywords = ["builder", "macro", "derive", "constructor", "setter"]

edition    = "2021"
homepage   = "https://bon-rs.com/"
license    = "MIT OR Apache-2.0"
repository = "https://github.com/elastio/bon"

# This MSRV was chosen because Rust supports mixing generic type and const
# parameters only starting with this version. We require this feature for the
# cases when the builder is generated for a function or struct that uses const
# generics. The generated builder **always** appends a generic type parameter
# for the type state at the end of the Builder generics list. So for functions
# or structs that use const generics the generated builder will have a generic
# type parameter after the const generic parameter, which is only supported
# starting from Rust 1.59.0.
rust-version = "1.59.0"

[package.metadata.docs.rs]
all-features = true

# Generate clickable links in the source code view in the docs
rustdoc-args = ["--generate-link-to-definition"]

# We don't need the docs to be built for every first-tier target.
# This crate is not platform-specific.
targets = ["x86_64-unknown-linux-gnu"]

[lints]
workspace = true

[dependencies]
# The version of the macro crate is pinned to a specific one because the code
# generated by the macros uses private APIs from the runtime crate that are not
# guarded by semver.
bon-macros  = { path = "../bon-macros", version = "=2.3.0" }
rustversion = "1"

[dev-dependencies]
# Using a bit older version that supports our MSRV
expect-test = "1.4.1"

# Using a bit older version that supports our MSRV
tokio = { version = "1.29.1", features = ["macros", "rt-multi-thread"] }

# Using a bit older version that supports our MSRV
trybuild = "1.0.89"

[features]
alloc   = []
default = ["std"]
std     = ["alloc"]

# Opts in to the higher MSRV 1.79.0. In this version, Rust stabilized the syntax
# for bounds in associated type position which can be used to make bounds on generic
# associated types implied. See the release announcement for more:
# https://blog.rust-lang.org/2024/06/13/Rust-1.79.0.htmlbounds-in-associated-type-position
#
# This feature is useful for the trait `IsComplete` generated by the builder macros.
# When this feature is enabled, the builder macros use the new syntax for bounds in
# associated type position, which enables implied `IsSet` bounds for the type state
# of required members.
#
# To understand how this can be used consider the following example:
#
# ```rust
# #[derive(bon::Builder)]
# struct Example {
#     a: u32,
#     b: Option<u32>,
# }
#
# use example_builder::{IsUnset, IsComplete};
#
# impl<State: example_builder::State> ExampleBuilder<State> {
#     fn build_with_default_b(self) -> Example
#     where
#         State: IsComplete,
#         State::B: IsUnset,
#     {
#         self.b(42).build()
#     }
# }
# ```
#
# This code wouldn't compile without this feature enabled, because `State: IsComplete`
# wouldn't automatically imply `State::A: IsSet`, so the builder type state returned
# after `self.b()` doesn't imply that the member `a` is set, and thus `build()`
# can't be called.
#
# Huge thanks to @harudagondi for suggesting the name of this cargo feature!
implied-bounds = ["bon-macros/implied-bounds"]

# üî¨ Experimental! There may be breaking changes to this feature between *minor* releases,
# however, compatibility within patch releases is guaranteed though.
#
# This feature enables the #[builder(overwritable)] attribute that can be used to
# allow overwriting already set fields in the builder. Without this attribute it's
# not possible to set the same member twice in the builder, a compile error would be
# generated.
#
# See more info at https://bon-rs.com/reference/builder/top-level/overwritable.
#
# We are considering stabilizing this attribute if you have a use for it. Please leave
# a üëç reaction under the issue https://github.com/elastio/bon/issues/149 if you need
# this attribute. It would also be cool if you could leave a comment under that issue
# describing your use case for it.
experimental-overwritable = ["bon-macros/experimental-overwritable"]
