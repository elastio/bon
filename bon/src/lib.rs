#![doc = include_str!("../README.md")]

pub use bon_macros::*;

/// Symbols used by macros. They are not stable and are considered an implementation detail.
#[doc(hidden)]
pub mod private;

/// Same as [`std::vec!`] but converts each element with [`Into::into()`].
///
/// **WARNING:** it's not recommended to import this macro into scope. Reference it
/// using the full path (`bon::vec![]`) to avoid confusion with the [`std::vec!`] macro.
///
/// A good example of the use case for this macro is when you want to create a
/// [`Vec<String>`] where part of the items are hardcoded string literals of type
/// `&str` and the other part is made of dynamic [`String`] values.
///
/// ```
/// fn convert_media(input_extension: &str, output_extension: &str) -> std::io::Result<()> {
///     let ffmpeg_args: Vec<String> = bon::vec![
///         "-i",
///         format!("input.{input_extension}"),
///         "-y",
///         format!("output.{output_extension}"),
///     ];
///
///     std::process::Command::new("ffmpeg").args(ffmpeg_args).output()?;
///
///     Ok(())
/// }
/// ```
///
/// This macro doesn't support `vec![expr; N]` syntax, since it's simpler to
/// just write `vec![expr.into(); N]` using [`std::vec!`] instead.
#[macro_export]
macro_rules! vec {
    () => (::std::vec::Vec::new());
    ($($item:expr),+ $(,)?) => (::std::vec![$(::core::convert::Into::into($item)),+ ]);
}

/// Creates a fixed-size array literal where each element is converted with [`Into::into()`]
/// into the target type.
///
/// You'll probably need a hint for the target type of items in the array if the
/// compiler can't infer it from its usage.
///
/// This is similar in spirit to the [`bon::vec!`] macro, but it's for arrays.
/// See [`bon::vec!`] docs for details.
///
/// Same example as in [`bon::vec!`], but using this macro. It works with array
/// as well because [`Command::args`] accepts any value that implements [`IntoIterator`]:
///
/// ```
/// fn convert_media(input_extension: &str, output_extension: &str) -> std::io::Result<()> {
///     let ffmpeg_args: [String; 4] = bon::arr![
///         "-i",
///         format!("input.{input_extension}"),
///         "-y",
///         format!("output.{output_extension}"),
///     ];
///
///     std::process::Command::new("ffmpeg").args(ffmpeg_args).output()?;
///
///     Ok(())
/// }
/// ```
///
/// This macro doesn't support `[expr; N]` syntax, since it's simpler to
/// just write `[expr.into(); N]` instead.
///
/// [`Command::args`]: std::process::Command::args
/// [`bon::vec!`]: crate::vec
#[macro_export]
macro_rules! arr {
    () => ([]);
    ($($item:expr),+ $(,)?) => ([$(::core::convert::Into::into($item)),+]);
}

#[cfg(test)]
mod tests {
    #[test]
    fn arr_smoke() {
        let actual: [String; 3] = crate::arr!["foo", "bar", "baz"];
        assert_eq!(actual, ["foo", "bar", "baz"]);

        let actual: [String; 0] = crate::arr![];
        assert!(actual.is_empty());
    }

    #[test]
    fn vec_smoke() {
        let actual: Vec<String> = crate::vec!["foo", "bar", "baz"];
        assert_eq!(actual, ["foo", "bar", "baz"]);

        let actual: Vec<String> = crate::vec![];
        assert!(actual.is_empty());
    }
}
