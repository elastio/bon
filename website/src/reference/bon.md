# `#[bon]`

This is a companion macro for [`builder`]. You should place it on top of the `impl` block
where you want to define methods with the [`builder`] macro. It provides
the necessary context to the [`builder`] macros on top of the functions
inside of the `impl` block. You'll get compile errors without that context.

For the examples of the usage of this macro and the reason why it's needed, see this paragraph in [the overview](../guide/overview#builder-for-an-associated-method).

[`builder`]: ./builder
