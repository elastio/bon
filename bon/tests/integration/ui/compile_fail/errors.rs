use bon::{builder, Builder};
use std::collections::{BTreeMap, BTreeSet};

fn main() {
    let map: BTreeMap<String, String> = bon::map! {
        "Hello": "Blackjack",
        "Hello": "Littlepip",
    };

    let set: BTreeSet<String> = bon::set!["mintals", "guns", "mintals", "roses"];

    #[derive(Builder)]
    struct SkipGeneratesNoSetter {
        #[builder(skip)]
        x: u32,

        #[builder(skip = 4)]
        y: u32,
    }

    SkipGeneratesNoSetter::builder().x(42).build();
    SkipGeneratesNoSetter::builder().y(42).build();

    #[derive(Builder)]
    struct Example {
        x: u32,
        y: u32,

        #[builder(name = renamed)]
        z: u32,
    }

    // Test error message about missing members
    let _ = Example::builder().x(1).build();

    // Test error message about repeated setter calls
    let _ = Example::builder().y(1).y(2);

    {
        type OpaqueOption<T> = Option<T>;

        #[derive(Builder)]
        struct Sut {
            arg1: OpaqueOption<u32>,
        }

        let _ = Sut::builder().build();
    }
}

#[derive(Builder)]
struct TupleStruct(u32, u32);

#[builder]
fn destructuring((x, y): (u32, u32)) {
    let _ = x;
    let _ = y;
}

#[builder]
fn unnecessary_into_false(#[builder(into = false)] _x: u32) {}

#[builder(on(String, into))]
fn unnecessary_into(#[builder(into)] _x: String) {}

#[builder(on(&dyn std::fmt::Debug, into))]
fn invalid_type_pattern() {}

#[builder(on(fn(#[attr] a: u32), into))]
fn attrs_in_on_type_pattern() {}

#[builder(on)]
fn incomplete_on() {}

#[builder(on())]
fn incomplete_on2() {}

#[builder(on(_))]
fn incomplete_on3() {}

#[builder(on(_,))]
fn incomplete_on4() {}

#[derive(Builder)]
#[builder(start_fn())]
struct EmptyStartFn {}

#[derive(Builder)]
struct ConflictingAttrs {
    #[builder(skip, into)]
    x: u32,
}

#[derive(Builder)]
struct ConflictingAttrs2 {
    #[builder(skip, name = bar)]
    x: u32,
}

#[derive(Builder)]
struct ConflictingAttrs3 {
    #[builder(skip, default = 42)]
    z: u32,
}

#[builder]
fn skip_on_fn_is_unsupporetd(
    #[builder(skip)] _x: u32,
    #[builder(skip = "skip".to_owned())] _y: String,
    #[builder(skip = vec![42])] _z: Vec<u32>,
) {
}

#[derive(Builder)]
struct TupleStructsAreUnsupported(u32, u32);

#[builder]
enum EnumsAreUnsupportedWithAttr {}

#[derive(Builder)]
enum EnumsAreUnsupportedWithDerive {}

#[builder]
fn destructuring_in_fn_is_unsupported((_, _): (u32, u32)) {}

#[builder]
#[must_use]
#[must_use]
fn double_must_use() {}

#[builder]
struct BuilderProcMacroAttrOnAStruct {}
