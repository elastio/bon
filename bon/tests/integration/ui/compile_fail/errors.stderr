error: duplicate key in the map
 --> tests/integration/ui/compile_fail/errors.rs:6:9
  |
6 |         "Hello": "Blackjack",
  |         ^^^^^^^

error: duplicate key in the map
 --> tests/integration/ui/compile_fail/errors.rs:7:9
  |
7 |         "Hello": "Littlepip",
  |         ^^^^^^^

error: duplicate value in the set
  --> tests/integration/ui/compile_fail/errors.rs:10:43
   |
10 |     let set: BTreeSet<String> = bon::set!["mintals", "guns", "mintals", "roses"];
   |                                           ^^^^^^^^^

error: duplicate value in the set
  --> tests/integration/ui/compile_fail/errors.rs:10:62
   |
10 |     let set: BTreeSet<String> = bon::set!["mintals", "guns", "mintals", "roses"];
   |                                                              ^^^^^^^^^

error: Only structs with named fields are supported
  --> tests/integration/ui/compile_fail/errors.rs:52:1
   |
52 | struct TupleStruct(u32, u32);
   | ^^^^^^

error: use a simple `identifier: type` syntax for the function argument; destructuring patterns in arguments aren't supported by the `#[builder]`
  --> tests/integration/ui/compile_fail/errors.rs:55:18
   |
55 | fn destructuring((x, y): (u32, u32)) {
   |                  ^^^^^^

error: Unexpected type `bool`
  --> tests/integration/ui/compile_fail/errors.rs:61:44
   |
61 | fn unnecessary_into_false(#[builder(into = false)] _x: u32) {}
   |                                            ^^^^^

error: this `#[builder(into)]` attribute is redundant, because `into` is already implied for this member via the `#[builder(on(...))]` at the top of the function
  --> tests/integration/ui/compile_fail/errors.rs:64:31
   |
64 | fn unnecessary_into(#[builder(into)] _x: String) {}
   |                               ^^^^

error: This syntax is not supported in type patterns yet. If you have a use case for this, please open an issue at https://github.com/elastio/bon/issues.
  --> tests/integration/ui/compile_fail/errors.rs:66:15
   |
66 | #[builder(on(&dyn std::fmt::Debug, into))]
   |               ^^^

error: nested attributes are not allowed in the type pattern of #[builder(on(type_pattern, ...))]
  --> tests/integration/ui/compile_fail/errors.rs:69:17
   |
69 | #[builder(on(fn(#[attr] a: u32), into))]
   |                 ^

error: Expected an attribute of form `on(type_pattern, ...)`
  --> tests/integration/ui/compile_fail/errors.rs:72:11
   |
72 | #[builder(on)]
   |           ^^

error: unexpected end of input, expected one of: `for`, parentheses, `fn`, `unsafe`, `extern`, identifier, `::`, `<`, `dyn`, square brackets, `*`, `&`, `!`, `impl`, `_`, lifetime
  --> tests/integration/ui/compile_fail/errors.rs:75:1
   |
75 | #[builder(on())]
   | ^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected `,`
  --> tests/integration/ui/compile_fail/errors.rs:78:1
   |
78 | #[builder(on(_))]
   | ^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: this #[builder(on(type_pattern, ...))] contains no options to override the default behavior for the selected setters like `into`, so it does nothing
  --> tests/integration/ui/compile_fail/errors.rs:81:1
   |
81 | #[builder(on(_,))]
   | ^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected at least one parameter in parentheses
  --> tests/integration/ui/compile_fail/errors.rs:85:11
   |
85 | #[builder(start_fn())]
   |           ^^^^^^^^

error: `skip` attribute can't be specified together with `into`
  --> tests/integration/ui/compile_fail/errors.rs:90:15
   |
90 |     #[builder(skip, into)]
   |               ^^^^

error: `skip` attribute can't be specified together with `name`
  --> tests/integration/ui/compile_fail/errors.rs:96:15
   |
96 |     #[builder(skip, name = bar)]
   |               ^^^^

error: `skip` attribute can't be specified with `default` attribute; if you wanted to specify a value for the member, then use the following syntax instead `#[builder(skip = value)]`
   --> tests/integration/ui/compile_fail/errors.rs:102:15
    |
102 |     #[builder(skip, default = 42)]
    |               ^^^^

error: `skip` attribute is not supported on function arguments. Use a local variable instead.
   --> tests/integration/ui/compile_fail/errors.rs:108:15
    |
108 |     #[builder(skip)] _x: u32,
    |               ^^^^

error: Only structs with named fields are supported
   --> tests/integration/ui/compile_fail/errors.rs:115:1
    |
115 | struct TupleStructsAreUnsupported(u32, u32);
    | ^^^^^^

error: only `fn` items are supported by the `#[bon::builder]` attribute
   --> tests/integration/ui/compile_fail/errors.rs:117:1
    |
117 | #[builder]
    | ^^^^^^^^^^
    |
    = note: this error originates in the attribute macro `builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: only `struct` items are supported by the `#[derive(bon::Builder)]` attribute
   --> tests/integration/ui/compile_fail/errors.rs:120:10
    |
120 | #[derive(Builder)]
    |          ^^^^^^^
    |
    = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: use a simple `identifier: type` syntax for the function argument; destructuring patterns in arguments aren't supported by the `#[builder]`
   --> tests/integration/ui/compile_fail/errors.rs:124:39
    |
124 | fn destructuring_in_fn_is_unsupported((_, _): (u32, u32)) {}
    |                                       ^^^^^^

error: Found multiple #[must_use], but bon only works with exactly one (or less).
   --> tests/integration/ui/compile_fail/errors.rs:128:1
    |
128 | #[must_use]
    | ^

warning: unused attribute
   --> tests/integration/ui/compile_fail/errors.rs:128:1
    |
128 | #[must_use]
    | ^^^^^^^^^^^ help: remove this attribute
    |
note: attribute also specified here
   --> tests/integration/ui/compile_fail/errors.rs:127:1
    |
127 | #[must_use]
    | ^^^^^^^^^^^
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: `#[warn(unused_attributes)]` on by default

warning: use of deprecated module `bon::private::deprecations::builder_attribute_on_a_struct`: #[bon::builder] on top of a struct is deprecated; use `#[derive(bon::Builder)]` instead; see more details at https://elastio.github.io/bon/blog/bon-builder-v2-2-release#derive-builder-syntax-for-structs
   --> tests/integration/ui/compile_fail/errors.rs:131:1
    |
131 | #[builder]
    | ^^^^^^^^^^
    |
    = note: `#[warn(deprecated)]` on by default
    = note: this warning originates in the attribute macro `builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `x` found for struct `SkipGeneratesNoSetterBuilder` in the current scope
  --> tests/integration/ui/compile_fail/errors.rs:21:38
   |
12 |     #[derive(Builder)]
   |              ------- method `x` not found for this struct
...
21 |     SkipGeneratesNoSetter::builder().x(42).build();
   |                                      ^ method not found in `SkipGeneratesNoSetterBuilder`

error[E0599]: no method named `y` found for struct `SkipGeneratesNoSetterBuilder` in the current scope
  --> tests/integration/ui/compile_fail/errors.rs:22:38
   |
12 |     #[derive(Builder)]
   |              ------- method `y` not found for this struct
...
22 |     SkipGeneratesNoSetter::builder().y(42).build();
   |                                      ^ method not found in `SkipGeneratesNoSetterBuilder`

error[E0277]: can't finish building yet; the member `ExampleBuilder__y` was not set
  --> tests/integration/ui/compile_fail/errors.rs:34:37
   |
34 |     let _ = Example::builder().x(1).build();
   |                                     ^^^^^ the member `ExampleBuilder__y` was not set
   |
   = help: the trait `IntoSet<u32, ExampleBuilder__y>` is not implemented for `Unset<Required>`
   = help: the trait `IntoSet<Option<_>, ExampleBuilder__y>` is implemented for `Unset<Optional>`
note: required by a bound in `ExampleBuilder::<(__X, __Y, __Z)>::build`
  --> tests/integration/ui/compile_fail/errors.rs:24:14
   |
24 |     #[derive(Builder)]
   |              ^^^^^^^ required by this bound in `ExampleBuilder::<(__X, __Y, __Z)>::build`
25 |     struct Example {
   |            ------- required by a bound in this associated function
   = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: can't finish building yet; the member `ExampleBuilder__renamed` was not set
  --> tests/integration/ui/compile_fail/errors.rs:34:37
   |
34 |     let _ = Example::builder().x(1).build();
   |                                     ^^^^^ the member `ExampleBuilder__renamed` was not set
   |
   = help: the trait `IntoSet<u32, ExampleBuilder__renamed>` is not implemented for `Unset<Required>`
   = help: the trait `IntoSet<Option<_>, ExampleBuilder__renamed>` is implemented for `Unset<Optional>`
note: required by a bound in `ExampleBuilder::<(__X, __Y, __Z)>::build`
  --> tests/integration/ui/compile_fail/errors.rs:24:14
   |
24 |     #[derive(Builder)]
   |              ^^^^^^^ required by this bound in `ExampleBuilder::<(__X, __Y, __Z)>::build`
25 |     struct Example {
   |            ------- required by a bound in this associated function
   = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: can't set the same member twice
  --> tests/integration/ui/compile_fail/errors.rs:37:37
   |
37 |     let _ = Example::builder().y(1).y(2);
   |                                     ^ this member was already set
   |
   = help: the trait `IsUnset` is not implemented for `Set<u32>`
   = help: the trait `IsUnset` is implemented for `Unset<T>`
note: required by a bound in `ExampleBuilder::<(__X, __Y, __Z)>::y`
  --> tests/integration/ui/compile_fail/errors.rs:24:14
   |
24 |     #[derive(Builder)]
   |              ^^^^^^^ required by this bound in `ExampleBuilder::<(__X, __Y, __Z)>::y`
...
27 |         y: u32,
   |         - required by a bound in this associated function
   = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: can't finish building yet; the member `SutBuilder__arg1` was not set
  --> tests/integration/ui/compile_fail/errors.rs:47:32
   |
47 |         let _ = Sut::builder().build();
   |                                ^^^^^ the member `SutBuilder__arg1` was not set
   |
   = help: the trait `IntoSet<Option<u32>, SutBuilder__arg1>` is not implemented for `Unset<Required>`
   = help: the trait `IntoSet<Option<u32>, SutBuilder__arg1>` is implemented for `Unset<Optional>`
   = help: for that trait implementation, expected `Optional`, found `Required`
note: required by a bound in `SutBuilder::<(__Arg1,)>::build`
  --> tests/integration/ui/compile_fail/errors.rs:42:18
   |
42 |         #[derive(Builder)]
   |                  ^^^^^^^ required by this bound in `SutBuilder::<(__Arg1,)>::build`
43 |         struct Sut {
   |                --- required by a bound in this associated function
   = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)
